### ISSUE TITLE
`napari.view_*` functions should be deprecated

### ISSUE THREAD

**Issue by jni** [napari/napari#7047]
## 🧰 Task

In #4693, we discussed that the `view_*` functions are awkward because they only return the viewer, where it would be more useful to return the viewer *and* the layer(s). This has now been implemented for images with `napari.imshow`, and we don't think we need them (yet) for the other layers, since napari is primarily an image viewer. We should deprecate the `view_` functions and make sure they don't appear in our docs or examples.

**Comment by TimMonko** [comment]
@jni Is this still the plan? Seems to be consensus still as goal after #7989 
If so, we can do the following
1. scipy sprints: change all usage of `view_*` to `napari.imshow` or 2 liner equivalent for other layers
2. add deprecation warning to view functions in 0.6.3
3. deprecate functions in 0.7.0

Seems clean, and the first point is great for sprints / first issues.

**Comment by jni** [comment]
Great plan @TimMonko! Thanks for organising!

**Comment by rossbar** [comment]
Hi napari-devs,

I just encountered the `FutureWarning` for `view_image` and wanted to take an opportunity to provide a bit of user feedback on the proposed (now in-place) deprecation!

Personally, I find the `napari.view_*` methods a very natural and practical interface for many use-cases, particularly for interactive, CLI-based usage. The vast majority of use-cases in my experience involve constructing layers interactively, for which having the ability to construct a viewer and populate a single layer (usually the "raw" image first) in a single line is quite convenient! My [typical usage](https://vanvalenlab.github.io/cellSAM/auto_examples/plot_yeaz.html#sphx-glr-auto-examples-plot-yeaz-py) looks something like:

```python
>>> nim = napari.view_image(img_data, ...)  # Create the viewer and populate the first layer all at once
>>> ... analyze: e.g. segment, extract key-points, etc.
>>> nim.add_labels(<mask>)
>>> nim.add_points(<keypoints>)
>>> # etc.
```

Now, AIUI there are two ways (one existing, one relatively new) to get what I expect - either create the Viewer explicitly:

```python
>>> nim = napari.Viewer()
>>> nim.add_image(...)
```

or with the newer `imshow`:

```python
>>> nim, _ = napari.imshow(img)
```

For the record I think these are both reasonable, but have some (admittedly extremely minor) drawbacks to the `view_*` API:
 1. The explicit instantiation of the `Viewer` class adds an additional line in files and feels clunkier to type interactively
 2. The multiple returns of `imshow` necessitates ignoring arguments (e.g. `nim, _ = napari.imshow`) if you're not particularly interested in the layer itself (quite common in my experience). Another minor (IMO) wart is the difference in return pattern with `matplotlib`'s imshow, which has a single return (the `AxisImage` object)[^1]

[^1]: I realize napari is quite different than matplotlib and the project is by no means beholden to the API decisions of others; nevertheless users will undoubtedly make associations due to the name!

Anyways - I just wanted to provide some feedback on the proposed API changes as a downstream user!  Of course I understand the depth of discussion that's typically involved in API changes, and fully recognize that updating our code/usage patterns is by no means an onerous task. I just wanted to float a bit of "practicality-beats-purity"-based advocacy for the existing `napari.view_*` functions!

Finally - apologies if I'm raising this discussion in the wrong place --- I'm happy to move it to a different issue/PR if there's a better place for it!

**Comment by psobolewskiPhD** [comment]
Thanks for the feedback @rossbar 
I think this is a perfectly fine place for the feedback! 

Personally, I've been a heavy `view_image` user -- it was my go-to -- so i've gone through exactly that process as you and am in total agreement. I've defaulted to the two line approach as `imshow` doesn't really resonate with me -- i've never been a heavy matplotlib user.

Unfortunately, I think the proverbial ship has sailed and the view methods will be removed -- but never say never, if there is more feedback about this, we might revert the deprecation? maybe? 😬  
However, would it be helpful for you if we pushed back the actual removal to a later release of napari (currently 0.7.0), just to give you more time to update any examples or scripts?

**Comment by TimMonko** [comment]
@rossbar, first of all, thank you so greatly for this feedback! 

It's great to hear what resonates with you both, as I must say I've always defaulted to `viewer = napari.Viewer()`. 
I think in my mind it makes sense that "I make the viewer, and then do things to the viewer" rather than having a magic function that does both! Note, for that reason I don't really have a strong feeling towards `view_*` or `imshow`. 

In fact, I often don't return the layer, and almost always find myself doing `viewer.layers[n]` to get back my layer. I think my perspective is that the code is "GUI in words". 

That's to say this is all very helpful context. I'm especially curious @jni thoughts on this. 
A good point is that not only does imshow return two objects, but it actually returns a `viewer, (layers)` which can be even more confusing if multiple layers are added with the command... 

Anyways, the machinery for the`view_*` functions to me looks pretty "simple" all things considered, so I don't know maybe we can put it in "it exists, but not actively maintained mode" and just keep it until it breaks? maybe not the best policy... Curious too what @brisvag and @Czaki think on something like just keeping the functions as we move towards napari-lite.

**Comment by Czaki** [comment]
Which data type that you use? Only `view_image` or also other layer types? 

What did you think about interface:

```
v = Viewer(layers=[(data, {**kwargs}, 'image')])
```

Where layers may accept `Layer` or `LayerDataTuple`.

**Comment by psobolewskiPhD** [comment]
I only ever used `view_image`, because I always start with an image 🤷 
I think your suggestion is definitely viable, more as an alternative to `viewer = Viewer(); viewer.add ...` than to `napari.view_...` It's certainly for the advanced user, rather than a newcomer. I at least would not put that as the get started with napari snippet....

**Comment by Czaki** [comment]
why 
```
v = napari.imshow(data, ...)[0]
```
Is it not enough? you mean that adding `[0]` is not convininent? 
Or is it just naming nad `view_image` may be a wrapper around `imshow`?

**Comment by psobolewskiPhD** [comment]
It's fine for me, I've already adjusted my workflows and workshops.
The comment from Ross https://github.com/napari/napari/issues/7047#issuecomment-3150614958 raises questions about what is clear, consistent, and easy to understand, particularly for someone who is not deeply embedded in napari.

But to answer your question, at a glance, if I wasn't already well familiar with imshow usage, I would expect:
`v = napari.imshow(data, ...)[0]`
to give me a viewer with a single layer from the image data, e.g. a single channel. Obviously that's wrong, but at a quick glance that's what that notation would suggest to me.

**Comment by TimMonko** [comment]
> It's fine for me, I've already adjusted my workflows and workshops. The comment from Ross [#7047 (comment)](https://github.com/napari/napari/issues/7047#issuecomment-3150614958) raises questions about what is clear, consistent, and easy to understand, particularly for someone who is not deeply embedded in napari.

To me this is the crux of the conversation --> what message is sent by our functions, their naming, and the conventions in python to the field. 

1. `viewer = napari.view_image()` feels a bit magic to me, since it's confusing I'm returning the viewing object and not the image, BUT I totally am sympathetic to the use case for Peter and Ross. I semantically read this as "view an image and return it"
2. `viewer, layers = napari.imshow()` feels like a good napari-only sweet spot, but this suffers from not following convention (i.e. matplotlib which is models) and requires some inherent knowledge of the function (whether thats via docs or signature inspection) to know its return types. I also would read this semantically as "show an image and return it"
3. `viewer = napari.Viewer(layers=LayerDataTuple, ...)` as @Czaki writes is actually a perfect sweet spot for me (as a one liner), because it semantically says to me "return a viewer object with these layers added (and any other args to the viewer)". But, requiring a LayerDataTuple is probably not ideal for a newcomer. This is currently not possible though, correct?

At any rate I would definitely see `napari.imshow()[0]` and think it would return the first layer of the LayerList / the first image of a series of images.

 Again, this comes down to things being different in napari-land, what our functions do convey, and what they could convey. 

It's for this reason, that if `viewer = napari.view_image()` makes sense to people as returning a viewer, then that's great.

**Comment by Czaki** [comment]
> But to answer your question, at a glance, if I wasn't already well familiar with imshow usage, I would expect:
> `v = napari.imshow(data, ...)[0]`
> to give me a viewer with a single layer from the image data, e.g. a single channel. Obviously that's wrong, but at a quick glance that's what that notation would suggest to me.

We may change `imshow` to return namedtuple and allow `viewer = napari.imshow().viewer`

**Comment by rossbar** [comment]
Apologies for the slow response (traveling)

>  However, would it be helpful for you if we pushed back the actual removal to a later release of napari (currently 0.7.0), just to give you more time to update any examples or scripts?

No need to push dates/versions - updating everything would not really be that big of a maintenance burden.

> Anyways, the machinery for theview_* functions to me looks pretty "simple" all things considered, so I don't know maybe we can put it in "it exists, but not actively maintained mode" and just keep it until it breaks?

This is kind of the direction I was coming from. I mostly wanted to provide feedback/concrete evidence of Hyrum's-law-lite - i.e. that you have *users* of this API even if it's not necessarily one that's recommended by the *developers*.

So *if* the primary motivation is simply API cleanup (i.e. getting rid of multiple explicit patterns to achieve the same thing), then that should be weighed against user convenience. Of course, IME at least, feedback from users about proposed changes is often extremely hard to come by. At least at the warnings stage - users are less shy about sharing feedback once their code breaks 🙃 . Anyways that's my only purpose here - to raise awareness that there is at least one (or seemingly at least 2 with @psobolewskiPhD ) users of this API for whom the proposed change would be a minor inconvenience.

**Comment by jni** [comment]
> We may change imshow to return namedtuple and allow viewer = napari.imshow().viewer

I like this suggestion. NamedTuples in general are a great idea.

@rossbar 

omg hi! 👋 

> My [typical usage](https://vanvalenlab.github.io/cellSAM/auto_examples/plot_yeaz.html#sphx-glr-auto-examples-plot-yeaz-py) looks something like

Just to reiterate others' question: do you ever use `view_*` for anything other than images?

> Another minor (IMO) wart is the difference in return pattern with matplotlib's imshow, which has a single return (the AxisImage object)

I would say that this is a false equivalence: imho the AxisImage object is more analogous to the Image layer object than to the Viewer, which is more like a Figure or Axes. And matplotlib has now for years discouraged the state machine API (`plt.imshow`) in favour of the OO API, so these days you would do (I think):

```python
fig, ax = plt.subplots()
axim = ax.imshow(image)
```

which is very much analogous (by design!) to:

```python
viewer = napari.Viewer()
layer = viewer.add_image(image)
```

Anyway, all of the above is just to add another perspective. I think it's certainly possible to revert the deprecation and keep the functions around more or less indefinitely — the burden is not huge — but I do want to make sure we explore the API space in case there's something we consider more "idiomatic" that you'd be happy with — like @Czaki's suggestions above.

Speaking of which, @Czaki:

> `v = Viewer(layers=[(data, {**kwargs}, 'image')])`

I think this is approaching what we will end up with with napari-lite and an everything-is-dataclasses approach. But I wouldn't want to implement it with layerdatatuples as they are dataclass-shaped but not dataclasses. I'd prefer, for example:

> `v = Viewer(layers=[{'data': data, 'type': 'image', **kwargs}])`

Brainwave: we could add a `version=` to layer specifications, which would allow us to define migrations just like we do for settings, and keep code working ~forever. Thoughts? (Maybe this warrants a new issue.)

**Comment by rossbar** [comment]
> Just to reiterate others' question: do you ever use view_* for anything other than images?

I'd say `view_image` is far and away most common in my personal experience, though there are cases where I've used `view_labels` to set up a viewer for e.g. qualitative comparison of image segmentation results.

> I would say that this is a false equivalence: imho the AxisImage object is more analogous to the Image layer object than to the Viewer, which is more like a Figure or Axes. And matplotlib has now for years discouraged the state machine API (plt.imshow) in favour of the OO API, so these days you would do (I think):

Definitely agreed here - my point was only about the different return types of `imshow`, i.e. one object vs. multiple. Honestly though I probably shouldn't even have mentioned it as I don't think it's relevant to the main point I wanted to raise re: the `view_` fns... just something that jumped out given the muscle memory from long-time use of maptlotlib!

> but I do want to make sure we explore the API space in case there's something we consider more "idiomatic" that you'd be happy with

When it comes to API design I'd assign negative weight to my own opinions 🙃  . FWIW I like the explicitness of `nim = napari.Viewer()` and I see plenty of value in all the existing and proposed ways of setting up a viewer and initializing layers. I just wanted to express how *convenient* I've personally found the `view_*` functions for my typical use case of interactive exploratory data analysis/viz!

**Comment by TimMonko** [comment]
Copying the start of a Zulip thread here for posterity: [#general > 0.7.0 Removal: &#96;napari.view_&#42;&#96; functions](https://napari.zulipchat.com/#narrow/channel/212875-general/topic/0.2E7.2E0.20Removal.3A.20.60napari.2Eview_*.60.20functions/with/543391600)

> Hi everyone, 0.7.0 is looking to be a pretty chunky release and will have some potentially impactful changes to how you use napari (we hope only in excellent ways)! You can generally follow along #**release>0.7.0** 
> 
> As we work towards release (planned in November), I'm going to be opening various threads to invite community feedback. This thread is the first of those! 
> 
> There has been a longtime plan to deprecate the `napari.view_*` functions (#4693 and #7047), where *only* a viewer is returned by effectively combining a `viewer = napari.Viewer()` and `viewer.add_*` calls. Now, this has either been replaced by `viewer, layer = napari.imshow()` for images, or more typically  instantiating the viewer and then adding the layer type of interest.
> 
> Do **you** have any thoughts about us removing this pattern? We already received excellent feedback from Ross Barnowski, and if you do care, it's an interesting read! https://github.com/napari/napari/issues/7047#issuecomment-3150614958
> 
> Recently, @**Wouter-Michiel Vierdag** pointed out that `numpydoc` is potentially a heavy dependency in #8322. When I went to work on this in #8334, I realized that most of the usage was for these crufty `view_*` functions, motivating further just ripping out these functions in #8337.
> 
> To me, we have the benefit of not only removing these unusual functions, but by also stepping towards removing `numpydoc` (#8338) the base napari install is 15% smaller (70MB decrease!) with 15 fewer dependencies! 
> 
> That's me giving *my opinion* (and you can see it in action with changes to tutorials in napari/docs#864), but we would love to hear from the community whether or not we should move forward with this deprecation. Some real difficulties of this implementation is that many tutorial materials, not just from napari, exist for years using the `viewer = napari.view_*` pattern.
